lib/private/AllConfig.php

```diff
@@ -304,6 +307,14 @@ public function setUserValue($userId, $appName, $key, $value, $preCondition = nu
 			}
 			$this->userCache[$userId][$appName][$key] = (string)$value;
 		}
+		$this->triggerUserValueChange($userId, $appName, $key, $value, $prevValue);
+	}
+
+	private function triggerUserValueChange($userId, $appId, $key, $value, $oldValue = null) {
+		if (\OC::$server instanceof \OCP\IServerContainer) {
+			$dispatcher = \OC::$server->get(IEventDispatcher::class);
```

Comment:
Cannot use DI for this as it is used in `base.php` over [here](https://github.com/nextcloud/server/blob/fd06f0c72de958855d535230b79583b30b9763b1/lib/base.php#L168)

Reply:
Is that the right link? The Allconfig constructor does not call setUserValue.

Reply:
@come-nc I believe I meant to say:

1. `AllConfig` is instantiated in `base.php` as `new \OC\AllConfig(new \OC\SystemConfig(self::$config))`
2. If I add `IEventListener $eventListener` in the constructor for DI in this class (the usual way), it could be problematic. I remember it causing errors when I first wrote this code, but maybe  I should re-test? 🤔

Reply:
Yes please, the event listener is available early so I would be surprised if this is called even earlier. Maybe for login time :thinking: 
If some value change call do not trigger the event we need to document that.

Reply:
@come-nc so if I use DI to set the event dispatcher, I will have to replace https://github.com/nextcloud/server/blob/fd06f0c72de958855d535230b79583b30b9763b1/lib/base.php#L168

with 

```php
new \OC\AllConfig(new \OC\SystemConfig(self::$config), new EventDispatcher(new SymfonyDispatcher(...), ))
```
This doesn't look like the best solution to me :|


I did test `new \OC\AllConfig(new \OC\SystemConfig(self::$config), Server::get(IEventDispatcher::class))` and this gets me the error: 

```
Error: Typed static property OC::$server must not be accessed before initialization in /var/www/html/lib/public/Server.php:55
Stack trace:
#0 /var/www/html/lib/base.php(174): OCP\Server::get('OCP\\EventDispat...')
```

Could you advise?

Reply:
My bad, AllConfig really is used really early, before the dispatcher can be injected.
So keep your solution, simple use the non-deprecated version:
```suggestion
			$dispatcher = \OCP\Server::get(IEventDispatcher::class);
```